File: Languages/cpp/low-level-design-lld/design-patterns/AbstractFactory.cpp
#include <iostream>
using namespace std;

// Define Abstract products
class Button {
  public:
    virtual void render() = 0;
    virtual ~Button() {}
};

class Checkbox {
  public:
    virtual void render() = 0;
    virtual ~Checkbox() {}
};

// Define Concrete Products
class WindowsButton : public Button {
  public:
    void render() override {
      cout << "Rendering Windows Button" << endl;
    }
};

class MacButton : public Button {
  public:
    void render() override {
      cout << "Rendering MAC button" << endl;
    }
};

class WindowsCheckbox : public Checkbox {
  public:
    void render() override {
      cout << "Rendering Windows checkbox " << endl;
    }
};

class MacCheckBox : public Checkbox {
  public:
    void render() override {
      cout << "Rendering MAC Checkbox" << endl;
    }
};

class GUIFactory {
  public:
    virtual Button* createButton() = 0;
    virtual Checkbox* createCheckbox() = 0;
    virtual ~GUIFactory() {}
};

class WindowsFactory : public GUIFactory {
  public:
    Button* createButton() override {
      return new WindowsButton();
    }

    Checkbox* createCheckbox() override {
      return new WindowsCheckbox();
    }
};

class MacFactory : public GUIFactory {
  public:
    Button* createButton() override {
      return new MacButton();
    }

    Checkbox* createCheckbox() override {
      return new MacCheckBox();
    }
};

class Application {
  private:
    GUIFactory* factory;
  public:
    Application(GUIFactory* f) : factory(f) {}

    void renderUI() {
      Button* btn = factory->createButton();
      Checkbox* cb = factory->createCheckbox();

      btn->render();
      cb->render();

      delete btn;
      delete cb;
    }
};

int main() {
  GUIFactory* factory = nullptr;

  factory = new WindowsFactory();
  Application app1(factory);
  app1.renderUI();
  delete factory;

  factory = new MacFactory();
  Application app2(factory);
  app2.renderUI();
  delete factory;

  return 0;
}

File: Languages/python/low-level-design-lld/design-patterns/abstract_factory.py
"""
Abstract Factory Pattern

Provides an interface for creating families of related objects without
specifying their concrete classes. Each factory produces a consistent
theme (Mac/Windows) of UI components.
"""

from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def render(self) -> str:
        pass


class Checkbox(ABC):
    @abstractmethod
    def render(self) -> str:
        pass


class MacButton(Button):
    def render(self) -> str:
        return "[ Mac Button ]"


class MacCheckbox(Checkbox):
    def render(self) -> str:
        return "[✓] Mac Checkbox"


class WindowsButton(Button):
    def render(self) -> str:
        return "[ Win Button ]"


class WindowsCheckbox(Checkbox):
    def render(self) -> str:
        return "[☑] Win Checkbox"


class UIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        pass

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        pass


class MacFactory(UIFactory):
    def create_button(self) -> Button:
        return MacButton()

    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()


class WindowsFactory(UIFactory):
    def create_button(self) -> Button:
        return WindowsButton()

    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()


def render_ui(factory: UIFactory) -> list[str]:
    """Client code works with any factory via the abstract interface."""
    button = factory.create_button()
    checkbox = factory.create_checkbox()
    return [button.render(), checkbox.render()]


if __name__ == "__main__":
    mac_ui = render_ui(MacFactory())
    assert mac_ui == ["[ Mac Button ]", "[✓] Mac Checkbox"]

    win_ui = render_ui(WindowsFactory())
    assert win_ui == ["[ Win Button ]", "[☑] Win Checkbox"]

    # Type consistency within a family
    mac = MacFactory()
    assert isinstance(mac.create_button(), MacButton)
    assert isinstance(mac.create_checkbox(), MacCheckbox)

    print("All tests passed!")
